crates/bevy_reflect/src/fields.rs

```diff
@@ -0,0 +1,79 @@
+use crate::Reflect;
+use std::any::TypeId;
+use std::borrow::{Borrow, Cow};
+
+/// The named field of a reflected struct
+#[derive(Clone, Debug)]
+pub struct NamedField {
+    name: Cow<'static, str>,
+    type_name: Cow<'static, str>,
+    type_id: TypeId,
+}
+
+impl NamedField {
+    pub fn new<T: Reflect>(name: &str) -> Self {
+        Self {
+            name: Cow::Owned(name.into()),
+            type_name: Cow::Owned(std::any::type_name::<T>().to_string()),
```

Comment:
This can be `Cow::Borrowed`. `std::any::type_name` returns a &'static str.

It'd also be ideal to have an alternative static version that uses static str that also borrow the name to avoid string allocation since the overwhelming vast majority of the cases generated from derive macros will also have &'static str fields names.

Reply:
> It'd also be ideal to have an alternative static version that uses static str that also borrow the name to avoid string allocation since the overwhelming vast majority of the cases generated from derive macros will also have &'static str fields names.

By this do you mean a completely separate type (like a `StaticNamedField` type)? Or a different constructor method like `const fn new_const(&str, &str, TypeId)`?

Reply:
Also, apparently `TypeId::of::<T>()` isn't `const`. That poses an issue if we wanted to make the methods themselves `const`.

Reply:
That's really unfortunate. Probably just going to have to do the boxed-slice optimization and call it a day.